src/theory/rewriter.cpp

```diff
@@ -81,15 +81,45 @@ struct RewriteStackElement {
   NodeBuilder<> builder;
 };
 
+RewriteResponse identityRewrite(RewriteEnvironment* re, TNode n)
+{
+  return RewriteResponse(REWRITE_DONE, n);
+}
+
 Node Rewriter::rewrite(TNode node) {
   if (node.getNumChildren() == 0)
   {
     // Nodes with zero children should never change via rewriting. We return
     // eagerly for the sake of efficiency here.
     return node;
   }
-  Rewriter& rewriter = getInstance();
-  return rewriter.rewriteTo(theoryOf(node), node);
+  return getInstance().rewriteTo(theoryOf(node), node);
+}
+
+void Rewriter::registerPreRewrite(
+    Kind k, std::function<RewriteResponse(RewriteEnvironment*, TNode)> fn)
+{
```

Comment:
What if k is EQUAL here? Should we add an assertion `Assert(k != EQUAL) << "Register pre-rewrites for EQUAL with registerPreRewriteEqual."`

Reply:
Why do we have two different functions to register a pre- and post-rewrite anyways?

Reply:
Huh, I somehow missed your comment @mpreiner. That's a good point, will do. @aniemetz some rewriters perform different rewrites in pre- and post-rewrites (e.g. the BV rewriter).

Reply:
No, what I meant is, why do we have a different function to register EQ and the rest?

Reply:
Ah because equalities are rewritten by all theories (depending on the operands) whereas other kinds are only associated with a single theory.

Reply:
We could still have a special handling in the register function for equalities no? I still don't get why we need 2 functions for this.

Reply:
It's more convenient this way. Otherwise, we would have to pass the theory to all rewrites, which is overkill.